File: src/main/scala/net/michalsitko/RequestLevel.scala
package net.michalsitko

import java.net.InetSocketAddress

import akka.actor.ActorSystem
import akka.http.scaladsl.model.HttpRequest
import akka.http.scaladsl.settings.{ClientConnectionSettings, ConnectionPoolSettings}
import akka.http.scaladsl.{ClientTransport, Http}
import akka.stream.ActorMaterializer
import net.michalsitko.utils.ResultLogger

import scala.concurrent.Future

object RequestLevel extends AnyRef with ResultLogger {
  def main(args: Array[String]): Unit = {
    implicit val system = ActorSystem()
    implicit val materializer = ActorMaterializer()
    implicit val ec = system.dispatcher

    val proxySettings = new InetSocketAddress("localhost", 8888)
    val transport = ClientTransport.proxy(None, proxySettings, ClientConnectionSettings(system))
//    val transport = ClientTransport.TCP(None, ClientConnectionSettings(system))

    val connectionPoolSettings = ConnectionPoolSettings(system).withTransport(transport)

    def singleRequest(uri: String): Future[Unit] = {
      val responseF = Http().singleRequest(HttpRequest(uri = uri), Http().defaultClientHttpsContext, connectionPoolSettings)
      logResult(responseF)
    }

    val done = for {
      _ <- singleRequest("https://github.com/akka/akka")
      _ <- singleRequest("https://twitter.com")
      _ <- singleRequest("https://www.scala-lang.org/resources/css/prettify.css")
      _ <- singleRequest("https://www.scala-lang.org/xyzxyz")
    } yield ()

    done.andThen {
      case _ =>
        Http().shutdownAllConnectionPools()
        materializer.shutdown()
        system.terminate()
    }
  }

}

File: src/main/java/net/michalsitko/java/RequestLevel.java
package net.michalsitko.java;

import akka.actor.ActorSystem;
import akka.http.javadsl.ClientTransport;
import akka.http.javadsl.Http;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.model.HttpResponse;
import akka.http.javadsl.model.Uri;
import akka.http.javadsl.settings.ClientConnectionSettings;
import akka.http.javadsl.settings.ConnectionPoolSettings;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;

import java.net.InetSocketAddress;
import java.util.Optional;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;

class RequestLevelClient {

    private final ActorSystem system;
    private final Materializer materializer;

    public RequestLevelClient(ActorSystem sys, Materializer mat) {
        system = sys;
        materializer = mat;
    }

    public ActorSystem getSystem() {
        return system;
    }

    public Materializer getMaterializer() {
        return materializer;
    }

    public void close() {
        System.out.println("Shutting down client");
        Http.get(system).shutdownAllConnectionPools().whenComplete((s, f) -> system.terminate());
    }

    private Uri uri = Uri.create("https://www.scala-lang.org/resources/css/prettify.css");

    public <U> CompletionStage<U> requestLevelFutureBased(Function<HttpResponse, CompletionStage<U>> responseHandler) {
        InetSocketAddress addr = new InetSocketAddress("localhost", 8888);
        ClientTransport clientTransport = ClientTransport.proxy(Optional.empty(), addr, ClientConnectionSettings.create(system));
        ConnectionPoolSettings settings = ConnectionPoolSettings.create(system).withTransport(clientTransport);
        return Http.get(system)
                .singleRequest(HttpRequest.create().withUri(uri), Http.get(system).defaultClientHttpsContext(), settings, system.log(), materializer)
                .thenComposeAsync(responseHandler);
    }



}

public class RequestLevel {
    public static void main(String[] args) {
        System.out.println("Starting...");
        final ActorSystem system = ActorSystem.create();
        final Materializer materializer = ActorMaterializer.create(system);

        RequestLevelClient client = new RequestLevelClient(system, materializer);
        client.requestLevelFutureBased(success -> {
                    System.out.println("some result");
                    return success.entity()
                            .getDataBytes()
                            .runForeach(byteString -> System.out.println("Result: " + byteString.utf8String()),
                                    client.getMaterializer());
                }
        ).whenComplete((success, throwable) -> client.close());
    }
}
